### PR TITLE
demonstrate VAP restricting action per-node

### PR DESCRIPTION
**Pull request by deads2k**
Add test showing how validating admission policy and SA token node claims can be used to enforce per-node mutation.

/sig auth
/kind cleanup

```release-note
NONE
```

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
Hello @deads2k  ! This PR has not been updated in some time, so I'd like to check what's the status. The code freeze is starting `02:00 UTC Wednesday 10th July 2024` (about `4 weeks` from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for `1.31`, is it still planned for this release?

**Comment by deads2k** [comment]
> As the PR is tagged for `1.31`, is it still planned for this release?

Yes, I've asked @enj for a review

**Comment by Vyom-Yadav** [comment]
Hello @deads2k @aramase,

This PR hasn't been updated in a while. What's the current status?

Reminder: Code freeze starts `02:00 UTC Wednesday 24th July 2024 / 19:00 PDT Tuesday 23rd July 2024` (about 1 week from now). Please make sure the PR has both `lgtm` and `approved` labels before the code freeze.

**Review by pohly — COMMENTED** [review]
Thanks @deads2k for this example. I managed to come up with a similar test case for DRA in https://github.com/kubernetes/kubernetes/pull/125488 based on this PR, see "DRA e2e: restrict DRA driver permissions with VAP" if you are interested.

While going through this PR I left some suggestions and questions.

File: test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "only-allow-name-matching-node-configmaps"
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+      - apiGroups:   [""]
+        apiVersions: ["v1"]
+        # notice you cannot restrict DELETEs.  Your per-node actor should not have deletion powers.
```
Comment (liggitt) [inline test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml:11]: why is this comment here? deletes can be restricted...

Comment (pohly) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml:11]: I can confirm that DELETEs get restricted exactly like CREATEs and UPDATEs - works for me.

Comment (pohly) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml:11]: Hmm, I spoke to soon. I hadn't updated by RBAC properly when running my test. Now that RBAC allows DELETEs, I am getting an error when evaluating the CEL expression:
```
message: 'resourceslices.resource.k8s.io "dra-fictional-node-slice" is forbidden:
                ValidatingAdmissionPolicy ''resourceslices-policy-dra-kubelet-plugin-dra-6661.k8s.io''
                with binding ''resourceslices-policy-dra-kubelet-plugin-dra-6661.k8s.io'' denied
                request: expression ''!variables.isKubeletPlugin || !variables.hasNodeName || has(object.spec.nodeName)
                && object.spec.nodeName == request.userInfo.extra["authentication.kubernetes.io/node-name"][0]''
                resulted in error: no such key: spec'
```

So it looks like VAP really cannot work for DELETEs because it doesn't make the object that is being deleted available in the CEL environment.

Perhaps it should, but currently it doesn't?

Comment (liggitt) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml:11]: object is not set in DELETE operations, oldObject is

Comment (liggitt) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml:11]: I'd suggest adding DELETE here and dropping the comment, then using the request.name attribute instead of pulling data out of the object, like this:

```
  matchConditions:
    - name: isRestrictedUser
      expression: request.userInfo.username == 'system:serviceaccount:e2e-ns:default'
  variables:
    - name: userNodeName
      expression: request.userInfo.extra[?'authentication.kubernetes.io/node-name'][0].orValue('')
  validations:
    - expression: variables.userNodeName == ''
      message: "no node association found for user, this user must run in a pod on a node"
    - expression: variables.userNodeName != request.name
      message: "this user may only modify configmaps that belong to the node the pod is running on"
```

File: test/e2e/auth/per_node_update.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+					},
+				},
+			},
+			Status: v1.PodStatus{},
+		})
+
+		ctx := context.Background()
+
+		// cleanup the ValidatingAdmissionPolicy.
+		defer func() {
```
[the remark this answers is not in the document]
Comment (deads2k) [reply test/e2e/auth/per_node_update.go]: > This is better done in a `ginkgo.DeferCleanup(func(ctx contextContxt) {...})`. That is a new context with different timeouts for cleanup operations.

Thanks!  I didn't know this existed.

File: test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml (hunk 1 above)
Comment (deads2k) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml]: > I'd suggest adding DELETE here and dropping the comment, then using the request.name attribute instead of pulling data out of the object.

Clever, though perhaps I'll open a followup showing how to use a field. The field is useful in cases where many instances of a resource can share a single node.

**Comment by deads2k** [comment]
I reviewed changes from https://github.com/deads2k/kubernetes/pull/10 and they looked very good.  Thanks!  Now squashed in.

File: test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml (hunk 1 above)
Comment (liggitt) [reply test/e2e/auth/e2edata/per_node_validatingadmissionpolicy.yaml]: Sure, you can pull a field value out for delete as well. Something like this hoists it into a variable later validations can reference succinctly with nice safe fallbacks to empty string:

```yaml
  variables:
  ...
    - name: objectNodeName
      expression: "request.operation == 'DELETE' ? oldObject : object).?whatever.nodeName.fieldPath.orValue('')"
```

File: test/e2e/framework/pod/exec_util.go [hunk 3]
```diff
@@ -68,8 +68,7 @@ func ExecWithOptionsContext(ctx context.Context, f *framework.Framework, options
 		Resource("pods").
 		Name(options.PodName).
 		Namespace(options.Namespace).
-		SubResource("exec").
-		Param("container", options.ContainerName)
```
Comment (pohly) [inline test/e2e/framework/pod/exec_util.go:72]: I fixed this while working on https://github.com/deads2k/kubernetes/pull/10: the container parameter in the URL gets added through v1.PodExecOptions and shouldn't be here (IMHO).

Might be better in a separate PR because it's not directly related. But it's small, so perhaps not worth it?

File: test/e2e/auth/per_node_update.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+		impersonatingClient, err := kubernetes.NewForConfig(impersonatingConfig)
+		framework.ExpectNoError(err)
+
+		noNodeAssociationMessage := "no node association found for user, this user must run in a pod on a node and ServiceAccountTokenPodNodeInfo must be enabled"
+		_, err = impersonatingClient.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, actualDisallowedConfigMap, metav1.CreateOptions{})
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(err.Error()).To(o.ContainSubstring(noNodeAssociationMessage))
+		_, err = impersonatingClient.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, actualAllowedConfigMap, metav1.CreateOptions{})
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(err.Error()).To(o.ContainSubstring(noNodeAssociationMessage))
```
Comment (pohly) [inline test/e2e/auth/per_node_update.go:213]: DELETE permission checking should also be tested.

**Comment by Vyom-Yadav** [comment]
/milestone clear
We are way past the code freeze. Please add the milestone for the next release if you'd like to target that.

File: test/e2e/auth/per_node_update.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		impersonatingConfig.Impersonate.Extra = map[string][]string{}
+		for k, v := range saUser.Status.UserInfo.Extra {
+			if k == "authentication.kubernetes.io/node-name" {
+				continue
+			}
+			currVal := append([]string{}, v...)
+			impersonatingConfig.Impersonate.Extra[k] = currVal
+		}
+		impersonatingClient, err := kubernetes.NewForConfig(impersonatingConfig)
+		framework.ExpectNoError(err)
```
Comment (enj) [inline test/e2e/auth/per_node_update.go:205]: It may be better to just do a token request directly with the cluster admin rest config.  Then you can just make requests as the SA without being bound to a node (the `make a kubeconfig with the token and confirm the kube-apiserver has the expected claims` logic above could be mostly re-used).

File: test/e2e/auth/per_node_update.go (hunk 4 above)
Comment (enj) [reply test/e2e/auth/per_node_update.go:213]: +1 on exercising the delete case as well.

**Review by enj — CHANGES_REQUESTED** [review]
Minor comments.

File: test/e2e/auth/per_node_update.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+		err = nodeScopedClient.CoreV1().ConfigMaps(f.Namespace.Name).Delete(ctx, actualDisallowedConfigMap.Name, metav1.DeleteOptions{})
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(err.Error()).To(o.ContainSubstring(disallowedMessage))
+
+		// recreate the allowedConfigMap and then do a delete collection
+		_, err = nodeScopedClient.CoreV1().ConfigMaps(f.Namespace.Name).Create(ctx, allowedConfigMap, metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+		err = nodeScopedClient.CoreV1().ConfigMaps(f.Namespace.Name).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+		o.Expect(err).To(o.HaveOccurred())
+		o.Expect(err.Error()).To(o.ContainSubstring(disallowedMessage))
```
Comment (deads2k) [inline test/e2e/auth/per_node_update.go:202]: DeleteCollection doesn't put name calls into each call to admission.  How interesting.... 

>    <string>: configmaps "Unknown" is forbidden: ValidatingAdmissionPolicy 'only-allow-name-matching-node-configmaps' with binding 'only-allow-name-matching-node-configmaps' denied request: expression 'variables.userNodeName == request.name' resulted in error: no such key: name
to contain substring
    <string>: this user running on node 'kind-worker' may not modify ConfigMap 'unlikely-node' because the name does not match the node name
In [It] at: k8s.io/kubernetes/test/e2e/auth/per_node_update.go:202 @ 08/19/24 19:52:31.287

File: test/e2e/auth/per_node_update.go (hunk 5 above)
Comment (deads2k) [reply test/e2e/auth/per_node_update.go:218]: Is this more than cosmetics?  Really think it's easier to reason about?

Comment (enj) [reply test/e2e/auth/per_node_update.go:218]: I don't feel strongly.

To me, a regular user would have an easier time testing such a VAP using `kubectl create token` which will always result in the correct user info instead of them trying to carefully impersonate it.

Maybe it would be best to use `kubectl` directly since this is an e2e and it would line up better with docs?

Comment (pohly) [reply test/e2e/auth/per_node_update.go:218]: I prefer to not depend on `kubectl` in an E2E test. The daemonset on the node is not going to use `kubectl` either.

**Comment by enj** [comment]
@deads2k looks like the delete collection test is failing, possibly because we return the first encountered error?

```
{ failed [FAILED] Expected
    <string>: configmaps "Unknown" is forbidden: ValidatingAdmissionPolicy 'only-allow-name-matching-node-configmaps' with binding 'only-allow-name-matching-node-configmaps' denied request: this user running on node 'kind-worker' may not modify ConfigMap 'kube-root-ca.crt' because the name does not match the node name
to contain substring
    <string>: this user running on node 'kind-worker' may not modify ConfigMap 'unlikely-node' because the name does not match the node name
In [It] at: k8s.io/kubernetes/test/e2e/auth/per_node_update.go:202 @ 08/22/24 21:50:44.55
}
```

File: test/e2e/auth/per_node_update.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		// Read manifests that we leave as yaml to make it clear to users how to these permissions are created.
+		// These will match future docs.
+		perNodeCheckValidatingAdmissionPolicy, err := e2eData.ReadFile("e2edata/per_node_validatingadmissionpolicy.yaml")
+		framework.ExpectNoError(err)
+		perNodeCheckValidatingAdmissionPolicyBinding, err := e2eData.ReadFile("e2edata/per_node_validatingadmissionpolicybinding.yaml")
+		framework.ExpectNoError(err)
+		admission := strings.ReplaceAll(string(perNodeCheckValidatingAdmissionPolicy), "e2e-ns", f.Namespace.Name)
+		admissionToCreate := readValidatingAdmissionPolicyV1OrDie([]byte(admission))
+		admissionBinding := strings.ReplaceAll(string(perNodeCheckValidatingAdmissionPolicyBinding), "e2e-ns", f.Namespace.Name)
+		admissionBindingToCreate := readValidatingAdmissionPolicyBindingV1OrDie([]byte(admissionBinding))
```
[the remark this answers is not in the document]
Comment (liggitt) [reply test/e2e/auth/per_node_update.go:64]: an incantation of the same resource builder kubectl uses to read objects from yaml and get the associated restmapper lookup could probably be used

File: test/e2e/auth/per_node_update.go (hunk 5 above)
Comment (liggitt) [reply test/e2e/auth/per_node_update.go:220]: +1 for not depending on kubectl, but I *would* probably create an actual pod-bound token with a tokenrequest directly, rather than using impersonation

File: test/e2e/feature/feature.go [hunk 8]
```diff
@@ -224,6 +224,11 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	NodeAuthenticator = framework.WithFeature(framework.ValidFeatures.Add("NodeAuthenticator"))
 
+	// Owner: sig-auth
+	// Marks tests that require a conforming implementation of
+	// Node claims for serviceaccounts
+	ServiceAccountTokenNodeBindingValidation = framework.WithFeature(framework.ValidFeatures.Add("ServiceAccountTokenNodeBindingValidation"))
```
[the remark this answers is not in the document]
Comment (liggitt) [reply test/e2e/feature/feature.go:231]: if your test only depends on presence of the node info based on the pod, you only need the `ServiceAccountTokenPodNodeInfo` gate

**Comment by deads2k** [comment]
```
I0904 18:03:29.169736 65026 per_node_update.go:221] Token: "eyJhbGciOiJSUzI1NiIsImtpZCI6ImduZ1RlOElmMWI1aktjbExTbE9OMjJJUzlmWDJBcHVrT292ZjVtYjFFU3cifQ.eyJhdWQiOlsiYXBpIl0sImV4cCI6MTcyNTQ3MzYwOSwiaWF0IjoxNzI1NDczMDA5LCJpc3MiOiJodHRwczovL2t1YmVybmV0ZXMuZGVmYXVsdC5zdmMuY2x1c3Rlci5sb2NhbCIsImp0aSI6IjIyNmZiOThlLTJhZTMtNGRjYi1hMWI5LTcwMTFjNjUwNmQ5YiIsImt1YmVybmV0ZXMuaW8iOnsibmFtZXNwYWNlIjoibm9kZS1hdXRobi00MjU1Iiwic2VydmljZWFjY291bnQiOnsibmFtZSI6ImRlZmF1bHQiLCJ1aWQiOiI0NWU5NjFlYi0zYzEzLTRhZGQtOGQyNC1mNzZhMzZmZGM1OGQifX0sIm5iZiI6MTcyNTQ3MzAwOSwic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50Om5vZGUtYXV0aG4tNDI1NTpkZWZhdWx0In0.duX4Wo_MpcTSlla99dJohPYMigR53DelJMhMTy6XTO-_m_tENk1gxzC1AUZs0Xns6B0GXxnyFwurijDi-561GckbGNzH5tZGTDljfEoLZZ5P9A_Vz65lE49ubUzI9s5FQUCNcAw6xH9CRjYk_Y20KPdhDQSm7vgNSbm1WMi9nuLd42Rpq16Q-5lHJi9ik75fsqH_v89GwaFfRU3DhebCsbqAuzCN8bHPCT6202wINm9JtRa1ZSuKLcKfWE0mQBpq7FWYQfc8O1E7Ug9ftwAZ2dvnXNsgIk_bCygP57WcmqcbvTnZqZxyLASH75Os6OgfzfehMks0c0w4W29A73gyng" expires at 2024-09-04 18:13:29 +0000 UTC
I0904 18:03:29.169921 65026 per_node_update.go:222] Unexpected error: 
    <*errors.StatusError | 0xc000fa3540>: 
    Unauthorized
    {
        ErrStatus: 
            code: 401
            message: Unauthorized
            metadata: {}
            reason: Unauthorized
            status: Failure,
    }
[FAILED] Unauthorized
```

huh, I wonder how I failed set to the token properly

File: staging/src/k8s.io/apiserver/pkg/audit/request.go [hunk 9]
```diff
@@ -53,6 +54,11 @@ func LogRequestMetadata(ctx context.Context, req *http.Request, requestReceivedT
 	ev.UserAgent = maybeTruncateUserAgent(req)
 	ev.Level = level
 
+	if ev.Annotations == nil {
+		ev.Annotations = make(map[string]string)
+	}
+	ev.Annotations["openshift.io/content-type"] = strings.Join(req.Header.Values("Content-Type"), ",")
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/audit/request.go:60]: Somehow I suspect that you did not mean to make this change? 😄
